store/postgres/src/relational.rs

```diff
@@ -518,21 +519,131 @@ impl Layout {
+            let entity = e.deserialize_with_layout::<Entity>(self, None)?;
+            let vid = ede.vid;
+            let ewt = EntityWithType {
+                entity_op,
+                entity_type,
+                entity,
+                vid,
+            };
+            Ok((ewt, block))
+        };
+
+        // The algorithm advances simultaneously entities from both lower_vec and upper_vec and tries
+        // to match entities that have entries in both vectors for a particular block. The match is
+        // successfull if an entry in one array has same values for the number of the block, entity
+        // name and the entity id. The comparison operation over the EntityDataExt fullfils that check.
+        // In addition to that, it also helps to order the elements so the algorith can detect if one
+        // side of the range exists and the other is missing. That way a creation and deletion are
+        // deduced. For immutable entities the entries in upper_vec are missing hence they are considered
+        // having a lower bound at particular block and upper bound at infinity.
+        while lower_now.is_some() || upper_now.is_some() {
+            let (ewt, block) = match (lower_now.is_some(), upper_now.is_some()) {
+                (true, true) => {
+                    match lower.cmp(&upper) {
+                        std::cmp::Ordering::Greater => {
+                            // we have upper bound at this block, but no lower bounds at the same block so it's deletion
+                            let (ewt, block) = transform(upper, EntitySubgraphOperation::Delete)?;
+                            // advance upper_vec pointer
+                            upper_now = upper_iter.next();
+                            upper = upper_now.unwrap_or(&EntityDataExt::default()).clone();
+                            (ewt, block)
+                        }
+                        std::cmp::Ordering::Less => {
+                            // we have lower bound at this block but no upper bound at the same block so its creation
+                            let (ewt, block) = transform(lower, EntitySubgraphOperation::Create)?;
+                            // advance lower_vec pointer
+                            lower_now = lower_iter.next();
+                            lower = lower_now.unwrap_or(&EntityDataExt::default()).clone();
+                            (ewt, block)
+                        }
+                        std::cmp::Ordering::Equal => {
+                            let (ewt, block) = transform(lower, EntitySubgraphOperation::Modify)?;
+                            // advance both lower_vec and upper_vec pointers
+                            lower_now = lower_iter.next();
+                            lower = lower_now.unwrap_or(&EntityDataExt::default()).clone();
+                            upper_now = upper_iter.next();
+                            upper = upper_now.unwrap_or(&EntityDataExt::default()).clone();
```

Comment:
What happens here if `lower_vec` contains multiple entries for the same entity? Advancing `upper_now` might move us to the next entity, and so we think that `lower_now` which has another version of the entity we just looked at will look like a create.

Reply:
I wonder whether a better strategy for all this might be to throw `upper_vec` and `lower_vec` into one vec, sort that by `(block, entity type, entity id)` and scan the resulting vec. You'd emit an `EntityWithType` if `block` is in the desired block range, and the operation will be determined with

* if previous entry had different type/id, it's a create
* if next entry has different type/id, it's a delete
* otherwise it's an update

That sorta requires a 3 element window to pass over the Vec, but since it's a Vec, that could be done with just using indices, like `vec[i-1]`, `vec[i]`, `vec[i+1]` with some special care for `i=0` and `i=vec.len()-1`, i.e. best to hide those intricacies in a helper functions.

Reply:
Not sure it it's possible to have multiple modifications of the entity in a block. That would entail a range of a type [10,10). In that case we might end up with wrong modification since we would have ranges like [3,10), [10,10), [10,15) and there is no guarantee that [10, 10) will be before [10,15). I guess if that's the case the DB request has to be extended with a 4. sorting criteria (for the lower_vec the end of the range, and for the upper_vec the start I believe)... Still if there are more than two changes in a block the order of the inner ones is not guaranteed to be correct, only the final result for a block...

Reply:
Not sure if merging the two vectors would make the algorithm simpler. It would entail having double entries for each entity that is modified. Also likely we would need a flag if it's a start or end of a range in order to deduce if there is a creation or deletion for single entries, also to avoid false creations at the first entry and false deletion at the last one. And to deduce the proper entry for update. Probably with that flag and 3 element window one can create such algorithm, but I believe it would have at least 6 checks if not more, where the current one has 5. Not to mention the additional complexity of the start/end boundaries in the helper function and the additional complexity needed for the SQL query.